File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/handoff.py
from __future__ import annotations

from pathlib import Path

from .common import read_json
from .prompt_context import compact_parent_handoff, render_structured_item


def shape_handoff(envelope: dict) -> dict:
    """
    Shape a compact handoff artifact from a completed delegate envelope.
    
    Handoff is durable cross-job context that is not session-dependent.
    It summarizes the normalized envelope without duplicating it, storing only
    what derived lineage jobs need to know about their parent.
    
    Invariant: handoff.json is the durable summarized context artifact, not a duplicate of operator_notes.
    """
    completion = envelope.get("completion", {})
    
    return {
        "ok": envelope.get("ok", False),
        "error_type": envelope.get("error_type"),
        "error_message": envelope.get("error_message"),
        "summary": completion.get("summary") or envelope.get("result") or "",
        "findings": envelope.get("findings", []),
        "open_risks": envelope.get("open_risks", []),
        "changed_files": envelope.get("changed_files", []),
        "diff_summary": envelope.get("diff_summary", []),
        "files_examined": envelope.get("files_examined", []),
        "test_commands": envelope.get("test_commands", []),
        "suggested_actions": envelope.get("suggested_actions", []),
        "task_packet_summary": envelope.get("task_packet_summary", {}),
        "execution_summary": {
            "duration_ms": envelope.get("duration_ms"),
            "num_turns": envelope.get("num_turns"),
            "total_cost_usd": envelope.get("total_cost_usd"),
        },
        "boundary_status": (envelope.get("boundary") or {}).get("status"),
        "verification_status": (envelope.get("verification") or {}).get("status"),
        "assistant_role": envelope.get("assistant_role"),
    }


def load_parent_handoff(job_path: str | Path) -> dict | None:
    """
    Load the parent handoff.json for a job.
    Returns None if the parent handoff does not exist.
    """
    from .common import artifact_paths
    
    paths = artifact_paths(Path(job_path))
    handoff_path = paths.get("handoff")
    if handoff_path is None or not handoff_path.exists():
        return None
    return read_json(handoff_path)


def render_parent_handoff(parent_handoff: dict | None) -> list[str]:
    if not parent_handoff:
        return []
    return render_structured_item(
        "Parent handoff:",
        compact_parent_handoff(parent_handoff),
    )

File: plugins/agent-skills/skills/claude-delegate/src/lib/handoff.ts
import { artifactPaths, readJson } from "./common.js";
import { compactParentHandoff, renderStructuredItem } from "./prompt-context.js";

export function shapeHandoff(
  envelope: Record<string, unknown>,
): Record<string, unknown> {
  const completion = (envelope.completion ?? {}) as Record<string, unknown>;

  return {
    ok: (envelope.ok as boolean) ?? false,
    error_type: envelope.error_type ?? null,
    error_message: envelope.error_message ?? null,
    summary:
      (completion.summary as string) ?? (envelope.result as string) ?? "",
    findings: (envelope.findings as unknown[]) ?? [],
    open_risks: (envelope.open_risks as unknown[]) ?? [],
    changed_files: (envelope.changed_files as unknown[]) ?? [],
    diff_summary: (envelope.diff_summary as unknown[]) ?? [],
    files_examined: (envelope.files_examined as unknown[]) ?? [],
    test_commands: (envelope.test_commands as unknown[]) ?? [],
    suggested_actions: (envelope.suggested_actions as unknown[]) ?? [],
    task_packet_summary: (envelope.task_packet_summary as Record<string, unknown>) ?? {},
    execution_summary: {
      duration_ms: envelope.duration_ms ?? null,
      num_turns: envelope.num_turns ?? null,
      total_cost_usd: envelope.total_cost_usd ?? null,
    },
    boundary_status:
      ((envelope.boundary as Record<string, unknown>) ?? {}).status ?? null,
    verification_status:
      ((envelope.verification as Record<string, unknown>) ?? {}).status ?? null,
    assistant_role: envelope.assistant_role ?? null,
  };
}

export function loadParentHandoff(
  jobPath: string,
): Record<string, unknown> | null {
  const paths = artifactPaths(jobPath);
  return readJson(paths.handoff);
}

export function renderParentHandoff(
  parentHandoff: Record<string, unknown> | null,
): string[] {
  if (!parentHandoff) return [];
  return renderStructuredItem(
    "Parent handoff:",
    compactParentHandoff(parentHandoff),
  );
}
